include/revng/Pipeline/Target.h

```diff
@@ -235,21 +212,73 @@ class TargetsList {
   template<typename... Args>
   void emplace_back(Args &&...A) {
     Contained.emplace_back(std::forward<Args>(A)...);
-    removeDuplicates();
+    llvm::sort(Contained);
+    Contained.erase(unique(Contained.begin(), Contained.end()),
+                    Contained.end());
   }
 
   void merge(const TargetsList &Other);
 
   void push_back(const Target &Target) {
     Contained.push_back(Target);
-    removeDuplicates();
+    llvm::sort(Contained);
+    Contained.erase(unique(Contained.begin(), Contained.end()),
+                    Contained.end());
   }
 
   template<typename... Args>
   auto erase(Args &&...A) {
     return Contained.erase(std::forward<Args>(A)...);
   }
 
+private:
+  struct Comp {
+    bool operator()(const Target &T, const Kind &K) const {
+      return &T.getKind() < &K;
+    }
+    bool operator()(const Kind &K, const Target &T) const {
+      return &K < &T.getKind();
+    }
+  };
+
+public:
+  TargetsList extracEntriesOfKind(const Kind &K) {
+    auto [b, e] = std::equal_range(begin(), end(), K, Comp());
+    TargetsList ToReturn;
+    copy(b, e, std::back_inserter(ToReturn));
+    Contained.erase(b, e);
+    return ToReturn;
+  }
+
+  TargetsList copyEntriesOfKind(const Kind &K) const {
+    auto [b, e] = std::equal_range(begin(), end(), K, Comp());
+    TargetsList ToReturn;
+    copy(b, e, std::back_inserter(ToReturn));
+    return ToReturn;
+  }
+
+  llvm::iterator_range<iterator> entriesOfKind(const Kind &K) {
+    auto [b, e] = std::equal_range(begin(), end(), K, Comp());
+    return llvm::make_range(b, e);
+  }
+
+  llvm::iterator_range<const_iterator> entriesOfKind(const Kind &K) const {
+    auto [b, e] = std::equal_range(begin(), end(), K, Comp());
+    return llvm::make_range(b, e);
+  }
+
+  llvm::SmallVector<const Kind *, 4> getContainedKinds() const {
+    llvm::SmallVector<const Kind *, 4> ToReturn;
+
+    auto Current = begin();
+    while (Current != end()) {
+      ToReturn.push_back(&Current->getKind());
+      Current = std::upper_bound(Current, end(), Current->getKind(), Comp());
+    }
```

Comment:
I feel like this has the same complexity as simply going through the list and adding if not the same kind as the previous one, but with extra steps. Can we make it dumber?

Reply:
if there is one kind for each target it has linear complexity yes. if there are thousands of isolated functions and only 1 or 2 kinds, this has logarithmic complexity.I would like to keep it since it used in the most intensive part of the pipeline, which is contract.cpp

Reply:
OK, but keep in mind that computing the set of things to compute will never be the bottleneck, so let's avoid early optimization.